python/perf-kernels/multreduce_matmul_kernel.py

```diff
@@ -1,45 +1,340 @@
+#!/usr/bin/env python
+
+# -*- coding: utf-8 -*-
+
+# Imports:
+# --------
+
+import argparse
+import sys
+from typing import Any, Optional
+
+import pytest
+import torch
+from torch import Tensor
+
 import triton
 import triton.language as tl
 
+# Input generation:
+# -----------------
+
+
+def gen_input(M: int, N: int, K: int, use_bias: bool, device: str = "cuda") -> tuple[Tensor, Tensor, Optional[Tensor]]:
+    assert M > 0, "M for input generation must be positive."
+    assert M <= 8, "M for input generation must be less or equal to 8."
+    assert N > 0, "N for input generation must be positive."
+    assert K > 0, "K for input generation must be positive."
+
+    a: Tensor = torch.randn((M, K), dtype=torch.float16, device=device)
+    b: Tensor = torch.randn((K, N), dtype=a.dtype, device=a.device)
+    bias: Optional[Tensor] = torch.randn(M, dtype=a.dtype, device=a.device) if use_bias else None
+
+    return a, b, bias
+
+
+# PyTorch GEMM:
+# -------------
+
+
+def torch_matmul(a: Tensor, b: Tensor, bias: Optional[Tensor]) -> Tensor:
+    c: Tensor = torch.matmul(a, b)
+    if bias is not None:
+        c += bias[:, None]
+    return c
+
+
+# Triton GEMM:
+# ------------
 
-# Kernel that implements GEMM with explicit multiply-reduce instructions for small block sizes.
-# Based on **tune_gemm** `matmul_kernel` from commit `cf44637` (see `triton-mlir` branch).
+
+def get_triton_autotune_configs() -> list[triton.Config]:
+    # yapf: disable
+    return [
+        triton.Config(
+            {
+                "BLOCK_SIZE_M": 1, "BLOCK_SIZE_N": 64, "BLOCK_SIZE_K": 256, "waves_per_eu": 0,
```

Comment:
Do you think any more config option will make sense here? Or will they add too much time to autotuning?

Reply:
Hi Rahul.

Let me explain my rationale behind these 3 autotuning configs. There's 4 target GEMM shapes, `tune_gemm.py` gave me 3 configs as the best ones, 2 target shapes share the same best config. So I added these 3 configs as autotuning options.

Probably we won't be choosing a good config if a shape other them the target ones is selected. How did you get your autotuning configs for RMSNorm and Softmax? I'm open to autotuning suggestion as well, feel free to suggest anything you think is appropriate. 

About autotuning time: To be honest, I didn't even worry about this. As far as I can see it's fast enough.

Thanks for your feedback.

Reply:
@brunomazzottiamd I just came up by making good guesses. I varied the num_warps and waves_per_eu. I do have a lot of auto-configs and for one kernel(softmax) the run takes a few mintues(but not too long). Also, I was running mostly shapes with lower number of rows.

Reply:
Hi @rahulbatra85. I'm going to rebase this and add more configs, varying `num_warps`. Do you have a study suggestion on `waves_per_eu`? It's always zero in `tune_gemm.py`, I want to better understand the meaning of this parameter.

Reply:
As I understand, it's a hint to the compiler to schedule waves_per_eu number of waves per execution unit(SIMD unit). More waves per eu will help with better utilization of GPU.

Reply:
I am not sure why it's set to 0 in tune_gemm.py. May be check with Shucai.

By the way, this config option is only valid for AMD GPUs

Reply:
I'm closing this discussion since it's very old and I added more configs to autotuning. I think any issue would be better covered in a new discussion.